File: src/com/huangyueran/spark/sql/RDD2DataFrameReflection.java
package com.huangyueran.spark.sql;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.sql.DataFrame;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.RowFactory;
import org.apache.spark.sql.SQLContext;
import org.apache.spark.sql.types.DataTypes;
import org.apache.spark.sql.types.StructField;
import org.apache.spark.sql.types.StructType;

/**
 * @category 动态创建schema元数据
 * @author huangyueran
 * @time 2017-7-24 13:58:59
 */
public class RDD2DataFrameReflection {
	public static void main(String[] args) {
		SparkConf conf = new SparkConf().setAppName("LoadSave").setMaster("local");
		JavaSparkContext sc = new JavaSparkContext(conf);
		SQLContext sqlContext = new SQLContext(sc);

		JavaRDD<String> lineRDD = sc.textFile("resources/people.txt");

		JavaRDD<Row> rowsRDD = lineRDD.map(new Function<String, Row>() {

			@Override
			public Row call(String line) throws Exception {
				String[] lineSplited = line.split(",");

				return RowFactory.create(lineSplited[0], Integer.valueOf(lineSplited[1]));
			}
		});

		// 动态构造元数据,这里用的动态创建元数据
		// 如果不确定有哪些列，这些列需要从数据库或配置文件中加载出来!!!!
		List<StructField> fields = new ArrayList<StructField>();
		fields.add(DataTypes.createStructField("name", DataTypes.StringType, true));
		fields.add(DataTypes.createStructField("age", DataTypes.IntegerType, true));

		StructType schema = DataTypes.createStructType(fields);

		// 根据表数据和元数据schema创建临时表
		DataFrame dataFrame = sqlContext.createDataFrame(rowsRDD, schema);
		dataFrame.registerTempTable("person");

		DataFrame personDataFrame = sqlContext.sql("select * from person");

		List<Row> list = personDataFrame.javaRDD().collect();

		// 一行记录
		for (Row r : list) {
			System.out.println(r);
		}

		sc.close();
	}

	public static class Person implements Serializable {
		private String name;
		private int age;

		public String getName() {
			return name;
		}

		public void setName(String name) {
			this.name = name;
		}

		public int getAge() {
			return age;
		}

		public void setAge(int age) {
			this.age = age;
		}

		@Override
		public String toString() {
			return "Person [name=" + name + ", age=" + age + "]";
		}

	}
}

File: src/main/scala/com/hyr/spark/sql/RDD2DataFrameReflection.scala
package com.hyr.spark.sql

import java.util
import java.util.{ArrayList, List}

import com.huangyueran.spark.utils.Constant
import com.hyr.spark.utils.SparkUtils
import org.apache.spark.sql.hive.HiveContext
import org.apache.spark.sql.types.{DataTypes, StructField, StructType}
import org.apache.spark.sql.{RowFactory, SparkSession}

/** *****************************************************************************
  *
  * @date 2019-08-29 15:33
  * @author: <a href=mailto:huangyr>黄跃然</a>
  * @Description:
  * *****************************************************************************/
object RDD2DataFrameReflection {

  def main(args: Array[String]): Unit = {
    val sparkContext = SparkUtils.getLocalSparkContext(RDD2DataFrameReflection.getClass)

    //val sparkSession = SparkSession.builder().appName("RDD2DataFrameReflection").enableHiveSupport.getOrCreate()
    val sparkSession = SparkSession.builder().appName("RDD2DataFrameReflection").getOrCreate()

    val lineRDD = sparkContext.textFile(Constant.LOCAL_FILE_PREX + "/data/resources/people.txt")
    //val lineRDD = sparkContext.textFile("/data/resources/people.txt")
    val rowsRDD = lineRDD.map(line => {
      val str = line.split(",")
      RowFactory.create(str(0), Integer.valueOf(str(1)))
    })
    rowsRDD

    // 动态构造元数据,这里用的动态创建元数据
    // 如果不确定有哪些列，这些列需要从数据库或配置文件中加载出来!!!!
    val fields = collection.mutable.ListBuffer[StructField]()
    fields += DataTypes.createStructField("name", DataTypes.StringType, true)
    fields += DataTypes.createStructField("age", DataTypes.IntegerType, true)

    val schema = DataTypes.createStructType(fields.toArray)
    schema.printTreeString() // 打印schema

    // 根据表数据和元数据schema创建临时表
    val dataSet = sparkSession.createDataFrame(rowsRDD, schema)
    dataSet.createOrReplaceTempView("person")

    // 通过sql查询
    val persons = sparkSession.sql("select * from person")
    persons.show
    val rows = persons.collect()

    for (s <- rows) {
      println(s)
    }

    sparkContext.stop

  }

}
